src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs

```diff
@@ -0,0 +1,422 @@
+        private MethodInfo _function;
+        private FunctionMetadataResolver _metadataResolver;
+        private readonly FunctionAssemblyLoader _assemblyLoader;
+        private Action _reloadScript;
+        private Action _restorePackages;
+
+        private static readonly string[] _watchedFileTypes = { ".cs", ".csx", ".dll", ".exe" };      
+
+        public CSharpFunctionInvoker(ScriptHost host, BindingMetadata trigger, FunctionMetadata metadata, bool omitInputParameter,
+            Collection<FunctionBinding> inputBindings, Collection<FunctionBinding> outputBindings, 
+            IFunctionEntryPointResolver functionEntryPointResolver, FunctionAssemblyLoader assemblyLoader)
+        {
+            _host = host;
+            _trigger = trigger;
+            _triggerParameterName = trigger.Name;
+            _omitInputParameter = omitInputParameter;
+            _inputBindings = inputBindings;
+            _outputBindings = outputBindings;
+            _functionMetadata = metadata;
+            _functionEntryPointResolver = functionEntryPointResolver;
+            _fileWatcher = InitializeFileWatcher(host);
+            _assemblyLoader = assemblyLoader;
+
+            if (_host.ScriptConfig.FileLoggingEnabled)
+            {
+                string logFilePath = Path.Combine(_host.ScriptConfig.RootLogPath, "Function", _functionMetadata.Name);
+                _fileTraceWriter = new FileTraceWriter(logFilePath, TraceLevel.Verbose);
+            }
+            else
+            {
+                _fileTraceWriter = NullTraceWriter.Instance;
+            }
+
+            _metadataResolver = new FunctionMetadataResolver(metadata, _fileTraceWriter);
+
+            _reloadScript = Debounce(ReloadScript);
+            _restorePackages = Debounce(RestorePackages);
+        }
+
+        private Action Debounce(Action targetAction, int milliseconds = 500)
+        {
+            Action<object> action = _ => targetAction();
+            action = action.Debounce(500);
+
+            return () => action(null);
+        }
+
+        private FileSystemWatcher InitializeFileWatcher(ScriptHost host)
+        {
+            FileSystemWatcher fileWatcher = null;
+            if (host.ScriptConfig.FileWatchingEnabled)
+            {
+                string functionDirectory = Path.GetDirectoryName(_functionMetadata.Source);
+                fileWatcher = new FileSystemWatcher(functionDirectory, "*.*")
+                {
+                    IncludeSubdirectories = true,
+                    EnableRaisingEvents = true
+                };
+                fileWatcher.Changed += OnScriptFileChanged;
+                fileWatcher.Created += OnScriptFileChanged;
+                fileWatcher.Deleted += OnScriptFileChanged;
+                fileWatcher.Renamed += OnScriptFileChanged;
+            }
+
+            return fileWatcher;
+        }
+
+        private void OnScriptFileChanged(object sender, FileSystemEventArgs e)
+        {
```

Comment:
We perform the "validation" compilation even if we're not loaded, so we always react to the file changes to get diagnostics.